File: opt/D-Prediction_and_Restriction.go
package main

import "fmt"

func main(){
	var n, k, r, s, p, ans int
	var T string
	fmt.Scan(&n, &k)
	fmt.Scan(&r, &s, &p)
	fmt.Scan(&T)

	memo := [] int {}
	for i := 0; i < n; i ++ {
		memo = append(memo, 0)
	}
	ans = 0
	for i := 0; i < n; i ++ {
		if i - k >= 0 {
			if T[i] != T[i - k] || memo[i - k] == 0 {
				if T[i] == 's' {
					ans += r
				} else if T[i] == 'p' {
					ans += s
				} else if T[i] == 'r' {
					ans += p
				}
				memo[i] = 1
			} 
			
		} else {
			if T[i] == 's' {
				ans += r
			} else if T[i] == 'p' {
				ans += s
			} else if T[i] == 'r' {
				ans += p
			}
			memo[i] = 1
		}
		

	}
	fmt.Println(ans)
	
	
}

File: D-Prediction_and_Restriction.py
n, k = map(int, input().split())
r, s, p = map(int,input().split())
T = input()
S = []
for i in range(n):
    if T[i] == 'r':
        S.append(p)
    elif T[i] == 's':
        S.append(r)
    else:
        S.append(s)

Memo = [0 for i in range(n)]
ans = 0
for i in range(n):
    if i - k >= 0:
        if S[i] != S[i - k] or Memo[i - k] == 0:
            ans += S[i]
            Memo[i] = 1
        
            
    else:
        ans += S[i]
        Memo[i] = 1
    
print(ans)
